cmd/kube-proxy/app/server.go

```diff
@@ -88,10 +89,25 @@ func (s *ProxyServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.HostnameOverride, "hostname-override", s.HostnameOverride, "If non-empty, will use this string as identification instead of the actual hostname.")
 	fs.BoolVar(&s.ForceUserspaceProxy, "legacy-userspace-proxy", true, "Use the legacy userspace proxy (instead of the pure iptables proxy).")
 	fs.DurationVar(&s.SyncPeriod, "iptables-sync-period", 5*time.Second, "How often iptables rules are refreshed (e.g. '5s', '1m', '2h22m').  Must be greater than 0.")
+	fs.BoolVar(&s.CleanupAndExit, "cleanup-iptables", false, "If true cleanup iptables rules and exit.")
 }
 
-// Run runs the specified ProxyServer.  This should never exit.
+// Run runs the specified ProxyServer.  This should never exit (unless CleanupAndExit is set).
 func (s *ProxyServer) Run(_ []string) error {
+	protocol := utiliptables.ProtocolIpv4
+	if s.BindAddress.To4() == nil {
+		protocol = utiliptables.ProtocolIpv6
+	}
+
+	// remove iptables rules and exit
+	if s.CleanupAndExit {
+		execer := exec.New()
+		ipt := utiliptables.New(execer, protocol)
+		userspace.TearDownIptablesProxierRules(ipt)
+		iptables.TearDownUserspaceProxierRules(ipt)
+		return nil
```

Comment:
I'm fine with a bool as long as the error messages remain logged.

On Thu, Aug 20, 2015 at 2:33 PM, Benjamin Elder <notifications@github.com>
wrote:

> In cmd/kube-proxy/app/server.go
> <https://github.com/kubernetes/kubernetes/pull/12981#discussion_r37564738>
> :
>
> >  func (s *ProxyServer) Run(_ []string) error {
> > +	protocol := utiliptables.ProtocolIpv4
> > +	if s.BindAddress.To4() == nil {
> > +		protocol = utiliptables.ProtocolIpv6
> > +	}
> > +
> > +	// remove iptables rules and exit
> > +	if s.CleanupAndExit {
> > +		execer := exec.New()
> > +		ipt := utiliptables.New(execer, protocol)
> > +		userspace.TearDownIptablesProxierRules(ipt)
> > +		iptables.TearDownUserspaceProxierRules(ipt)
> > +		return nil
>
> Right. So I need to find the k8s way to combine and return errors at the
> end (perhaps `[]error` ?). That or we could just return a `bool
> encounteredError` since we already log the errors in the method.
> … <#14f4c62f3799185b_>
> On Thu, Aug 20, 2015 at 2:30 PM, James DeFelice <notifications@github.com>
> wrote: In cmd/kube-proxy/app/server.go <
> https://github.com/kubernetes/kubernetes/pull/12981#discussion_r37564272>
> : > func (s *ProxyServer) Run(_ []string) error { > + protocol :=
> utiliptables.ProtocolIpv4 > + if s.BindAddress.To4() == nil { > + protocol
> = utiliptables.ProtocolIpv6 > + } > + > + // remove iptables rules and exit
> > + if s.CleanupAndExit { > + execer := exec.New() > + ipt :=
> utiliptables.New(execer, protocol) > +
> userspace.TearDownIptablesProxierRules(ipt) > +
> iptables.TearDownUserspaceProxierRules(ipt) > + return nil I like that they
> continue to tear down rules - I don't want that part to change. I just want
> the caller of the proc to know whether the teardown was clean or not. …
> <#14f4c60207fcdf36_> On Thu, Aug 20, 2015 at 2:22 PM, Benjamin Elder <
> notifications@github.com> wrote: In cmd/kube-proxy/app/server.go <
> https://github.com/kubernetes/kubernetes/pull/12981#discussion_r37563400>
> : > func (s *ProxyServer) Run(_ []string) error { > + protocol :=
> utiliptables.ProtocolIpv4 > + if s.BindAddress.To4() == nil { > + protocol
> = utiliptables.ProtocolIpv6 > + } > + > + // remove iptables rules and exit
> > + if s.CleanupAndExit { > + execer := exec.New() > + ipt :=
> utiliptables.New(execer, protocol) > +
> userspace.TearDownIptablesProxierRules(ipt) > +
> iptables.TearDownUserspaceProxierRules(ipt) > + return nil I'll do that,
> just need to find what the k8s idiom is for returning 0-n errors. We want
> to continue trying to tear down rules in TearDownIptablesProxierRules and
> TearDownUserspaceProxierRules even if one fails so right now they just
> glog.Errorf(...) errors instead of returning them. — Reply to this email
> directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/12981/files#r37563400>. —
> Reply to this email directly or view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/12981/files#r37564272>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12981/files#r37564738>.
>